### DISCUSSION TITLE
Linking to data via relationships

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#6252] [Help]
When I was using my own MYSQL database, if I set up a relationship, there was a column added similar to `fk_A_B`

So, If I created a view which showed me all the details of the job and the customer was `Customer x` I could use a link element and populate the customers name using the data provider info and link the customer to their account page using the URL `/customer/{{ fk_A_B }}`

The Budibase DB doesnt seem to add this column when setting up a relationship.

Is there a way to achieve the same outcome?

**Comment by melohagan** [comment]
Hi @PaulMcF1987 

I think I need to visualise this. Would you be able to share some screenshots and/or an app export please?

**Reply by PaulMcF1987** [reply]
Hi @melohagan hope your well and looking forward to the weekend.

So for example if I have a record page, for account contacts... the name of the company that they work for is displayed. 
Previously which using my own MYSQL DB I displayed the company name as a link and linked it using the fk column in the table (I am assuming that this means foreign key)

![image](https://user-images.githubusercontent.com/75368133/173024003-ce73d50a-7ea7-48d5-a6ec-a9c1d17ee377.png)

The above pic should show how I set up the link.

There doesnt seem to be an `fk` column with budibase DB.
I switched over thinking that using the internal DB might make things easier but it seems to make some things easier (Using attachments etc) but other things harder.

**Reply by melohagan** [reply]
Hope things are good with you too; I am glad it's Friday! 😄 

For the internal table, you have the relationship column type: https://docs.budibase.com/docs/relationships
It was always be an array (even if it only holds one relationship). 

So for example, if you have:
- One Billing -> One Account

Then the binding would look roughly like: `{{ Repeater.Billing.Account.0._id }}`

If you also shared some screenshots of your Billing and Account tables, I could give you the exact binding.

**Comment by PaulMcF1987** [comment]
@melohagan 

![image](https://user-images.githubusercontent.com/75368133/173028980-4467b154-afcc-4de6-a4e0-a2cab7208b3c.png)
![image](https://user-images.githubusercontent.com/75368133/173029163-d5e7bfad-6c04-4962-8f8a-3cbe9c0d5632.png)


Above is the headers for the contacts

Below is the company list which. each person in the above table will work for a company in the table belwo

![image](https://user-images.githubusercontent.com/75368133/173028869-ec70c79d-d482-4b75-9a36-624ce1e5ad46.png)
![image](https://user-images.githubusercontent.com/75368133/173029262-728f7b1a-cc1a-4118-99ca-df5aafe923fd.png)

**Reply by melohagan** [reply]
So the relationship is:
- One Billing -> One Company -> One Account ?

**Reply by PaulMcF1987** [reply]
@melohagan Im not sure of the wording used when setting it up... I am sure that one billing contact can only be linked to one company. but a company can have many billing contacts.

**Reply by PaulMcF1987** [reply]
@melohagan 
In the Billing table, the company choice is under the `company` column.
In the account Accounts table, the company name is also under the `Company` column. Budibase appears to have added a `Billing` column to link to the billing contact

**Reply by PaulMcF1987** [reply]
@melohagan 
I got it... I added a new data provider and chose the `relationships - repeater.billing.company`
and set the link to `/account/{{ New Data Provider.Rows.0._id }}

I suppose this then brings me to another question...

IS there something similar I can do to filter table views by customer?

For example, if I give a customer access to the app, I dont want them to be able to see all of the work I have done, only the work for them.

Preciously I have filtered using the `fk` column in the table

Kinda like
`if email equals X then 'fk' equals (number assigned in fk column for that account)`

**Reply by melohagan** [reply]
As you are using the Internal table, you could try using the auto-generated Created-By/Updated-By fields as the data source.
There's an example here: https://docs.budibase.com/docs/fetching-data#display-user-specific-data

**Reply by PaulMcF1987** [reply]
@melohagan 
I dont think that will work as different people work on accounts. I need a way to filter by company name

basically if my customers company if companya and his email address is x@companya.com
When he logs into the app, I only want him to be able to see jobs for Companya and all the rest of the jobs are filtered out.

**Reply by PaulMcF1987** [reply]
@melohagan 
I have tried adding another column then within css `display:none`

On chosing the custoemr from the dropdown, I have chose the onchange option to update the hidden box with the `field value`

This then puts a number like `ro_ta_4d803d1e95c4419f9862993051b72479_61424e5b60...................` into the box.

I then filtered the columns based on this number.

I am not sure how viable this is though... will the number stay the same? is there potential for issues doing it this way?

**Reply by melohagan** [reply]
Ah yes I see. If you are confident that the company name will be in the email address and consistent between users of that company, then there is a binding for the logged in users' email: `{{ Current User.email }}`

So assuming you want to filter on the company name, but from a relationship (i.e. the table you're filtering does not directly have the company name) then you'll need to add that column, called *CompanyName*, with a **static** Formula:
https://docs.budibase.com/docs/formula#fields-from-relationships

Then your filter will be something like
`CompanyName equals <JS Binding>`

**JavaScript**
```
return $("Current User.email").split('@')[1].split('.com')[0]
```

Although there's probably more email variations you'd need to account for.

**Reply by melohagan** [reply]
You can also add custom columns to the users table. So it might make sense to try and add the company name to the users table, maybe with a formula:
https://docs.budibase.com/docs/users-table#custom-columns

**Reply by PaulMcF1987** [reply]
@melohagan 
Not all of my customers will use their company name in the email address some might use the name of the parent group etc.
Which is why I need something consistent that will work for every customer.

for example... what I mentioned above...

I have added a new field so when I chose the customer, it populates the new field with something similar to `ro_ta_4d803d1e95c4419f9862993051b72479_61424e5b60...................`

I then used that to filter the table and add a `view`...

Then set a condition where currentuser.email = user@companya.com change data source to `The view I created`

I just feel that this is very messy and am unsure how well it will work. Is there any potential for errors to occur? Or data to get wrongly filtered?

**Reply by PaulMcF1987** [reply]
@melohagan 

> So assuming you want to filter on the company name, but from a relationship (i.e. the table you're filtering does not directly have the company name) then you'll need to add that column, called CompanyName, with a static Formula:

Seems to do the trick
